core/arch/x86/emit_utils.c

```diff
@@ -1363,8 +1363,11 @@ append_restore_simd_reg(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
         post_restore = INSTR_CREATE_label(dcontext);
         pre_avx512_restore = INSTR_CREATE_label(dcontext);
         APP(ilist,
-            INSTR_CREATE_cmp(dcontext, OPND_CREATE_ABSMEM(d_r_avx512_code_in_use, OPSZ_1),
-                             OPND_CREATE_INT8(0)));
+            INSTR_CREATE_cmp(
+                dcontext,
+                OPND_CREATE_ABSMEM(
+                    vmcode_get_executable_addr((byte *)d_r_avx512_code_in_use), OPSZ_1),
```

Comment:
We have the same problem in clients: w^x breaks DR_ALLOC_CACHE_REACHABLE.  Could you add an assert just like the one for dr_nonheap_alloc and add #3556 to the PR.  (Can test on client.alloc test disabling the dr_nonheap_alloc calls to reach DR_ALLOC_CACHE_REACHABLE).

Reply:
I've added the assertion in the same manner as in dr_nonheap_alloc. Resolving. Let me know if !ok.